protocol/buff/buff.go

```diff
@@ -0,0 +1,288 @@
+	b.payload = append(b.payload, tmp...)
+}
+
+// PushUUID writes a types.UUID to the buffer.
+func (b *Buff) PushUUID(val types.UUID) {
+	b.payload = append(b.payload, val[:]...)
+}
+
+// PushBytes writes a []byte to the buffer.
+func (b *Buff) PushBytes(val []byte) {
+	b.PushUint32(uint32(len(val)))
+	b.payload = append(b.payload, val...)
+}
+
+// PushString writes a string to the buffer.
+func (b *Buff) PushString(val string) {
+	b.PushUint32(uint32(len(val)))
+	b.payload = append(b.payload, val...)
+}
+
+// BeginBytes allocates space for `data_length` in the buffer.
+// May be called multiple times to create nested bytes blocks.
+// Calling EndBytes once for each BeginBytes call is required
+// before ending a message.
+// BeginBytes panics if BeginMessage was not called first.
+func (b *Buff) BeginBytes() {
+	if b.mPos <= 0 {
+		panic("cannot begin bytes: no current message")
+	}
+
+	b.bPos = append(b.bPos, len(b.payload))
+	b.payload = append(b.payload, 0, 0, 0, 0)
+}
+
+// EndBytes sets the `data_length` allocated by BeginBytes
+// to the number of bytes that were written since the last BeginBytes call.
+// EndBytes panics if BeginBytes was not called first.
+func (b *Buff) EndBytes() {
+	n := len(b.bPos)
+	if n < 1 {
+		panic("cannot end bytes: no bytes in progress")
+	}
+
+	p := b.bPos[n-1]
+	b.bPos = b.bPos[:n-1]
+	length := uint32(len(b.payload) - p - 4)
+	binary.BigEndian.PutUint32(b.payload[p:p+4], length)
+}
+
+// BeginMessage writes mType to the buffer
+// and allocates space for message length.
+// BeginMessage panics if the EndMessage was not called
+// for the previous message.
+func (b *Buff) BeginMessage(mType uint8) {
+	if b.mPos > 0 {
+		panic("cannot begin message: the previous message is not finished")
+	}
+
+	b.mPos = 1 + len(b.payload)
+	b.payload = append(b.payload, mType, 0, 0, 0, 0)
+}
+
+// EndMessage sets the `message_length` allocated by BeginMessage.
+// EndMessage panics if BeginMessage was not called first
+// or if BeginBytes was not followed by EndBytes.
+func (b *Buff) EndMessage() {
+	// todo check unfinished bytes
+	if b.mPos <= 0 {
+		panic("cannot end message: no current message")
+	}
+
+	if len(b.bPos) > 0 {
+		panic("cannot end message: bytes in progress")
+	}
+
+	length := uint32(len(b.payload) - b.mPos)
+	binary.BigEndian.PutUint32(b.payload[b.mPos:b.mPos+4], length)
+	b.mPos = 0
+}
+
+// Unwrap returns a pointer to the buffers *[]byte.
+func (b *Buff) Unwrap() *[]byte {
+	if b.mPos > 0 {
+		panic("cannot unwrap: the previous message is not finished")
+	}
+
+	return &b.payload
+}
+
+// Next returns true if the buffer is not fully read.
+func (b *Buff) Next() bool {
+	b.Finish()
+
+	if b.mPos < len(b.payload) {
+		if len(b.payload) < b.mPos+5 {
+			panic("buffer overread")
+		}
+
+		pos := 1 + int(binary.BigEndian.Uint32(b.payload[b.mPos+1:b.mPos+5]))
+
+		if len(b.payload) < b.mPos+pos {
+			panic("buffer overread")
+		}
+
+		b.Msg = b.payload[b.mPos+5 : b.mPos+pos]
```

Comment:
I assume we copy data here, is that right? Or this is like a "memory view"?

Reply:
slices in go are a view into an array they are [literally](https://golang.org/pkg/reflect/#SliceHeader)
```go
type SliceHeader struct {
    Data unsafe.Pointer
    Len  int
    Cap  int
}
```

Assigning to a slice only sets the three fields. It does not copy the underlying data.

Reply:
Awesome.